<a href="https://colab.research.google.com/github/shrutin0492/Deep-Learning-and-Computer-Vision-Lab/blob/main/4_BayesClassifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
#importing necessary libraries
import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer
from nltk import NaiveBayesClassifier
from nltk.classify import accuracy

# Download necessary resources
nltk.download('punkt')
nltk.download('stopwords')

# Preprocessing function
def preprocess(text):
    tokens = word_tokenize(text.lower())  # Tokenization and lowercase
    stop_words = set(stopwords.words('english'))
    tokens = [word for word in tokens if word.isalpha() and word not in stop_words]  # Removing stopwords and non-alphabetic tokens
    stemmer = PorterStemmer()
    tokens = [stemmer.stem(word) for word in tokens]  # Stemming
    return dict([(token, True) for token in tokens])

# Read dataset from a text file
dataset_file = "/content/drive/MyDrive/dlcv_lab/Exp4/dataset.txt"  # Path to your dataset file
dataset = []
with open(dataset_file, 'r') as file:
    for line in file:
        text, label = line.strip().split(",")
        dataset.append((text, label))

# Preprocess the dataset
preprocessed_dataset = [(preprocess(text), label) for text, label in dataset]

# Split data into training and testing sets
train_data = preprocessed_dataset[:90]
test_data = preprocessed_dataset[10:]

# Train the Naive Bayes Classifier
classifier = NaiveBayesClassifier.train(train_data)

# Test the classifier
print("Accuracy:", accuracy(classifier, test_data))

# Test a new text segment
text_to_classify = "The product exceeded my expectations"  # You can change this text to test different segments
preprocessed_text = preprocess(text_to_classify)
print("Classification:", classifier.classify(preprocessed_text))


Accuracy: 0.9333333333333333
Classification: positive


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


# Increasing Accuracy with Random Forest

In [29]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.feature_extraction import DictVectorizer


vectorizer = DictVectorizer()
X_train = vectorizer.fit_transform([text for text, label in train_data])
y_train = [label for text, label in train_data]
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(X_train, y_train)
X_test = vectorizer.transform([text for text, label in test_data])
y_test = [label for text, label in test_data]
accuracy = rf_classifier.score(X_test, y_test)
print("Accuracy:", accuracy)
def preprocess(text):
    tokens = word_tokenize(text.lower())
    stop_words = set(stopwords.words('english'))
    tokens = [word for word in tokens if word.isalpha() and word not in stop_words]
    stemmer = PorterStemmer()
    tokens = [stemmer.stem(word) for word in tokens]
    return {token: True for token in tokens}
text_to_classify = "The product exceeded my expectations"
preprocessed_text = preprocess(text_to_classify)
X_new = vectorizer.transform([preprocessed_text])
classification = rf_classifier.predict(X_new)[0]
print("Classification:", classification)

Accuracy: 0.9777777777777777
Classification: positive


Next Steps:
- accuracy >0.98
- add more preprocessing techniques
- use different classifiers